tests/virt_autotest/esxi_open_vm_tools.pm

```diff
@@ -0,0 +1,218 @@
+
+    reset_consoles;
+    select_console("sut");
+    assert_screen("grub2", 90);
+    send_key 'ret';
+    select_console("root-console", timeout => 180);
+
+    assert_script_run("ping -c3 openqa.suse.de");
+    assert_script_run("ping -c3 www.suse.com");
+}
+
+sub clock_sync_tests {
+    my ($vm_name, $vm_id, $vm_ip) = @_;
+    my ($g_datetime, $h_datetime, $g_difftime);
+
+    # Enable timesync service in guest VM
+    if (script_output("vmware-toolbox-cmd timesync status", proceed_on_failure => 1, timeout => 90) == "Disabled") {
+        $g_datetime = script_output("date -u +'\%Y-\%m-\%d \%H:\%M:\%S'");
+        record_info('Guest current time before syncing up', $g_datetime);
+        assert_script_run("vmware-toolbox-cmd timesync enable", timeout => 90);
+        $h_datetime = get_host_timestamp();
+        record_info('Host time', $h_datetime);
+        $g_datetime = script_output("date -u +'\%Y-\%m-\%d \%H:\%M:\%S'");
+        record_info('Guest time after syncing up with host', $g_datetime);
+    }
+
+    my $h_timesec = get_timelc($h_datetime);
+    my $g_timesec = get_timelc($g_datetime);
+    my $diff_sec  = int($g_timesec - $h_timesec);
+
+    record_info('Diff secs', $diff_sec);
+    if ($diff_sec <= 10) {
+        record_info("Clock synchronization is successful.");
+    } else {
+        die "Clock synchronization failed.";
+    }
+
+    select_console('svirt');
+
+    # Disable clock sync tests
+    if (esxi_vm_power_getstate($vm_id) =~ m/Powered on/) {
+        esxi_vm_power_shutdown($vm_id);
+        my $num = 6;
+        while ($num && console('svirt')->run_cmd("! ping -c1 $vm_ip")) {
+            sleep(10);
+            $num--;
+        }
+    }
+
+    disable_all_clock_synchronization($vm_name);
+
+    if (esxi_vm_power_getstate($vm_id) =~ m/Power off/) {
+        esxi_vm_power_on($vm_id);
+        my $num = 6;
+        while ($num && console('svirt')->run_cmd("ping -c1 $vm_ip")) {
+            sleep(10);
+            $num--;
+        }
+    }
+
+    reset_consoles;
+    select_console('sut');
+    assert_screen("grub2", 90);
+    send_key 'ret';
+    select_console("root-console", timeout => 180);
+
+    # Set the guest system time which different from the host
+    my $n_hrs = (sprintf "%02d", rand(24));
+    my $n_min = (sprintf "%02d", rand(59));
+    my $n_sec = (sprintf "%02d", rand(59));
+    $g_difftime = script_output("date -s $n_hrs:$n_min:$n_sec");
+    record_info('Set a time in guest that different from host', $g_difftime);
+
+    # Guest time will not be synced up with the hypervisor after time synchronization disabled
+    $g_datetime = script_output("date -u +'\%Y-\%m-\%d \%H:\%M:\%S'");
+    $h_datetime = get_host_timestamp();
+    $g_timesec  = get_timelc($g_datetime);
+    $h_timesec  = get_timelc($h_datetime);
+    $diff_sec   = int($g_timesec - $h_timesec);
+
+    record_info('Diff secs', $diff_sec);
+    if (abs($diff_sec) > 300) {
+        record_info("Clock synchronization was disabled.");
```

Comment:
I really like the above design that tests disabled clock sync. Can similar way (set differentiated time on guest, then start clock sync, then check the time diff with host)  be used to test enabled clock sync? IMHO, it will be better and more accurate. BTW is guest reboot necessary here?

Reply:
Yes, we can do that before enabling clock sync service.
From the official document, power off guest is the prerequisites before disabling clock sync. Referring to the link below.
https://docs.vmware.com/en/VMware-Tools/11.2.5/com.vmware.vsphere.vmwaretools.doc/GUID-678DF43E-5B20-41A6-B252-F2E13D1C1C49.html